wincrypto/src/srtp.rs

```diff
@@ -0,0 +1,316 @@
+                None,
+                &key,
+                0,
+            ))?;
+        }
+        Ok(Self(key_handle))
+    }
+}
+
+impl Drop for WinCryptoSrtpKey {
+    fn drop(&mut self) {
+        unsafe {
+            if let Err(e) = WinCryptoError::from_ntstatus(BCryptDestroyKey(self.0)) {
+                error!("Failed to destory crypto key: {}", e);
+            }
+        }
+    }
+}
+
+pub fn wincrypto_srtp_aes_128_ecb_round(
+    key: &WinCryptoSrtpKey,
+    input: &[u8],
+    output: &mut [u8],
+) -> Result<usize, WinCryptoError> {
+    unsafe {
+        // Run AES
+        let mut count = 0;
+        WinCryptoError::from_ntstatus(BCryptEncrypt(
+            key.0,
+            Some(input),
+            None,
+            None,
+            Some(output),
+            &mut count,
+            BCRYPT_BLOCK_PADDING,
+        ))?;
+        Ok(count as usize)
+    }
+}
+
+pub fn wincrypto_srtp_aes_128_cm(
+    key: &WinCryptoSrtpKey,
+    iv: &[u8],
+    input: &[u8],
+    output: &mut [u8],
+) -> Result<usize, WinCryptoError> {
+    unsafe {
+        // First, we'll make a copy of the IV with a countered as many times as needed into a new
+        // countered_iv.
+        let mut iv = iv.to_vec();
+        let mut countered_iv = [0u8; MAX_BUFFER_SIZE];
+        let mut offset = 0;
+        while offset <= input.len() {
+            let mut _count = 0;
+            let start = offset;
+            let end = offset + 16;
+            countered_iv[start..end].copy_from_slice(&iv);
+            offset += 16;
+            for idx in 0..16 {
+                let n = iv[15 - idx];
+                if n == 0xff {
+                    iv[15 - idx] = 0;
+                } else {
+                    iv[15 - idx] += 1;
+                    break;
+                }
+            }
+        }
+
+        // Now, we'll encrypt the countered IV. CNG can do this in-place, so we'll need a separate
+        // reference to the slice, but fool the borrow-checker, otherwise it won't like us passing
+        // the immutable and mutable reference to BCryptEncrypt.
+        let encrypted_countered_iv =
+            std::slice::from_raw_parts_mut(countered_iv.as_mut_ptr(), countered_iv.len());
+        let mut _count = 0;
+        WinCryptoError::from_ntstatus(BCryptEncrypt(
+            key.0,
+            Some(&countered_iv[..offset]),
+            None,
+            None,
+            Some(&mut encrypted_countered_iv[..offset]),
+            &mut _count,
+            BCRYPT_FLAGS(0),
+        ))?;
+
+        // XOR the intermediate_output with the input
+        for i in 0..input.len() {
+            output[i] = input[i] ^ encrypted_countered_iv[i];
+        }
+        Ok(input.len())
+    }
+}
+
+pub fn wincrypto_srtp_aead_aes_128_gcm_encrypt(
+    key: &WinCryptoSrtpKey,
+    iv: &[u8],
+    additional_auth_data: &[u8],
+    plain_text: &[u8],
+    cipher_text: &mut [u8],
+) -> Result<usize, WinCryptoError> {
+    unsafe {
```

Comment:
some places you are having everything in unsafe blocks and some places only the unsafe function I would be consistent here with one style

Reply:
Went through and tightened the unsafe blocks, to only encompass the unsafe calls. In a few places it's a few separate calls, but I tried to tighten up so it's clear what is "unsafe" and what is in theory safe.